### PR TITLE
Ability to send emails

### PR DESCRIPTION
**Pull request by Bowser65**
Emails are mjml backed in the /email folder. Those mjml files gets transpiled to HTML and pushed to /src/main/resources/email/html

### CODE REVIEW

File: src/main/resources/email/strings/en.properties [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+titles.password_reset=Reset your password
+
+### Contents
+# General
+contents.hi=Hi {{ pseudo }},
+contents.welcome=Welcome to Learndesk {{ pseudo }}!
+contents.paste=Or copy/paste this link in your browser:
+
+# Data harvest
+harvest.intro=You requested us the {{ request_date }} your Learndesk data archive, and we since then worked on baking it with a recipe of our grandma.
```
Comment (vinceh121) [inline src/main/resources/email/strings/en.properties]: On the {{ request_date }}, you requested your Learndesk data archive, and we've been baking it with a recipe of our grandma.

File: src/main/resources/email/strings/en.properties [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+harvest.tuto=The data archive is a zip file including all data stored on our backend. To open it, just extract all files from the zip archive, and open the <u>index.html</u> file with your favourite web browser.
+harvest.questions=If you have any question about the archive or you have concerns about your privacy on Learndesk, feel free to reach us at privacy@learndesk.app.
```
Comment (vinceh121) [inline src/main/resources/email/strings/en.properties]: If you have any questions or concerns about the archive or your privacy on Learndesk, feel free to reach us at privacy@learndesk.app.

File: src/main/resources/email/strings/en.properties [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+# Password Reset
+password_reset.there=It seems that you just asked us for a password recovery, so here we go. Click on the button below to reset your password.
+password_reset.ignore=If you didn't requested that password reset, then just ignore this email.
+password_reset.reset=Reset my password
+
+### Footer
+footer.reason.1=You received this email because you have a registered account on Learndesk
+footer.reason.2=You received this email because you've requested it
+footer.questions=Any question about this email ? Reach us at support@learndesk.app
```
Comment (vinceh121) [inline src/main/resources/email/strings/en.properties]: Any question about this email? Reach us at support@learndesk.app

File: src/main/kotlin/app/learndesk/Mail.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    private val locales: Map<Locale, Properties>
+
+    private val scheduler = Executors.newSingleThreadExecutor()!!
+
+    init {
+        // LOCALES
+        val enLocales = {}.javaClass.getResource("/email/strings/en.properties").openStream()
+        val en = Properties()
+        en.load(enLocales)
+        enLocales.close()
```
Comment (vinceh121) [inline src/main/kotlin/app/learndesk/Mail.kt]: Need to automate this for all locales

File: src/main/kotlin/app/learndesk/Mail.kt [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            scheduler.submit {
+                email.send()
+            }
+        } catch (e: Throwable) {
+            log.error("Failed to send email!", e)
+        }
+    }
+
+    private fun bakeMail(email: String, locale: Locale, variables: Map<String, Any>): Pair<String, String> {
+        val locales = locales[locale] ?: error("wtf dude")
```
Comment (vinceh121) [inline src/main/kotlin/app/learndesk/Mail.kt]: need a better error message

File: src/main/kotlin/app/learndesk/Mail.kt [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+    fun send(to: String, emailId: String, locale: Locale, variables: Map<String, Any> = emptyMap()) {
+        val mail = bakeMail(emailId, locale, variables)
+        try {
+            val email = HtmlEmail()
+            email.hostName = Learndesk.properties.getProperty("smtp.host")
+            email.subject = mail.first
+
+            email.addTo(to)
+            email.setCharset("UTF-8")
```
Comment (vinceh121) [inline src/main/kotlin/app/learndesk/Mail.kt]: Need to switch to UTF-16 if you want to be able to have non latin characters

File: src/main/kotlin/app/learndesk/Mail.kt [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            email.setFrom("noreply@learndesk.app")
+            scheduler.submit {
+                email.send()
+            }
+        } catch (e: Throwable) {
+            log.error("Failed to send email!", e)
+        }
+    }
+
+    private fun bakeMail(email: String, locale: Locale, variables: Map<String, Any>): Pair<String, String> {
```
Comment (vinceh121) [inline src/main/kotlin/app/learndesk/Mail.kt]: Unit tests needed

File: src/main/kotlin/app/learndesk/Mail.kt (hunk 5 above)
Comment (Bowser65) [reply src/main/kotlin/app/learndesk/Mail.kt]: Here we'll be sure that the locale we get exists as it'll get filtered before as the user won't be able to set a non-existing locale through the API. That error will never occur in real world situation but is here se `locales` is typed as `String` and not `String?`

File: src/main/kotlin/app/learndesk/Mail.kt (hunk 7 above)
Comment (Bowser65) [reply src/main/kotlin/app/learndesk/Mail.kt]: Units test will be done soon:tm:

**Pull request merged**